systems/framework/lcm_driven_loop.h

```diff
@@ -0,0 +1,227 @@
+  LcmDrivenLoop(drake::lcm::DrakeLcm* drake_lcm,
+                std::unique_ptr<drake::systems::Diagram<double>> diagram,
+                const drake::systems::LeafSystem<double>* lcm_parser,
+                std::vector<std::string> input_channels,
+                const std::string& active_channel,
+                const std::string& switch_channel)
+      : drake_lcm_(drake_lcm), lcm_parser_(lcm_parser) {
+    // Move simulator
+    if (!diagram->get_name().empty()) {
+      diagram_name_ = diagram->get_name();
+    }
+    diagram_ptr_ = diagram.get();
+    simulator_ =
+        std::make_unique<drake::systems::Simulator<double>>(std::move(diagram));
+
+    // Create subscriber for the switch (in the case of multi-input)
+    DRAKE_DEMAND(!input_channels.empty());
+    if (input_channels.size() > 1) {
+      DRAKE_DEMAND(!switch_channel.empty());
+      switch_sub_ = std::make_unique<drake::lcm::Subscriber<SwitchMessageType>>(
+          drake_lcm_, switch_channel);
+    }
+
+    // Create subscribers for inputs
+    for (const auto& name : input_channels) {
+      std::cout << "Constructing subscriber for " << name << std::endl;
+      name_to_input_sub_map_.insert(std::make_pair(
+          name, drake::lcm::Subscriber<InputMessageType>(drake_lcm_, name)));
+    }
+
+    // Make sure input_channels contains active_channel, and then set initial
+    // active channel
+    bool is_name_match = false;
+    for (const auto& name : input_channels) {
+      if (name.compare(active_channel) == 0) {
+        is_name_match = true;
+        break;
+      }
+    }
+    DRAKE_DEMAND(is_name_match);
+
+    active_channel_ = active_channel;
+  };
+
+  // Start simulating the diagram
+  void Simulate(double end_time = std::numeric_limits<double>::infinity()) {
+    // Get mutable contexts
+    auto& diagram_context = simulator_->get_mutable_context();
+    auto& lcm_parser_context = diagram_ptr_->GetMutableSubsystemContext(
+        *lcm_parser_, &diagram_context);
+
+    // Wait for the first message.
+    drake::log()->info("Waiting for first lcm input message");
+    LcmHandleSubscriptionsUntil(drake_lcm_, [&]() {
+      return name_to_input_sub_map_.at(active_channel_).count() > 0;
+    });
+
+    // Initialize the context based on the first message.
+    const double t0 =
+        name_to_input_sub_map_.at(active_channel_).message().utime * 1e-6;
+    diagram_context.SetTime(t0);
+    auto& input_value = lcm_parser_->get_input_port(0).FixValue(
+        &lcm_parser_context,
+        name_to_input_sub_map_.at(active_channel_).message());
+
+    // "Simulator" time
+    double time = 0;  // initialize the current time with 0
+    double message_time;
+
+    drake::log()->info(diagram_name_ + " started");
+    // Run the simulation until end_time
+    while (time < end_time) {
+      // Update the name of the active channel if there are multiple inputs
+      if (switch_sub_ != nullptr) {
```

Comment:
One (hopefully last) comment. For better readability, with the current structure, it feels like this code (executing the channel switch) should come after the message is received (rather than before).

I know it doesn't matter much, but I think it's easier to understand if the code looks like:
```
while true
   wait for message A or B
   if A, change channel
   if B, step simulation
```
rather than
```
while true
   if A, change channel
   wait for message A or B
   if B, step simulation
```

Reply:
Done. I also updated the logic of clearing messages, so now we clear the new channel messages during the loop when we just switch to a new channel, in case we read the old messages.